tests/test_json_schema.py

```diff
@@ -2692,73 +2683,202 @@ class Model(BaseModel):
         pet: Union[Cat, Dog, Lizard] = Field(..., discriminator='pet_type')
 
     assert Model.model_json_schema() == {
+        '$defs': {
+            'Cat': {
+                'properties': {'pet_type': {'const': 'cat', 'title': 'Pet Type'}},
+                'required': ['pet_type'],
```

Comment:
I think a human authored schema is a lot more likely to spec out `pet_type` a level up, to ensure it's always uniformly typed. I.e. I'd probably hand-write something like:

```
  '$defs': {
            'Cat': {'const': 'cat'},
            'Dog': {'const': 'dog'},
            'Lizard': {'enum': ['reptile', 'lizard']},
  },
  "properties": {
    "pet": {
      "required": ["pet_type"],
      "properties": {
        "pet_type": {
          "type": "string",
          "oneOf": {
            {"$ref": "#/$defs/Cat"},
            {"$ref": "#/$defs/Dog"},
            {"$ref": "#/$defs/Lizard"},
          }
       }
  }
  
```

(As I say whether that's a ton more difficult implementation-wise I haven't considered)

But what you have indeed looks valid to me.

Maybe one implementation note is you may consider in your test suite having a guard assertion that the schema you generated is valid (shameless plug obviously, but that's calling [`Draft202012Validator.check_schema(whatyouhavehere)`](https://python-jsonschema.readthedocs.io/en/stable/api/jsonschema/protocols/#jsonschema.protocols.Validator.check_schema) in my implementation).